File: jupyter2/answers/minc/minc/ml/minc/lib/codegen.ml
open Ast

let ( let* ) = Option.bind

(* report an error at the point it occurs and return None *)
let fail msg =
  let () = Printf.eprintf "codegen error: %s\n" msg in
  None

(* run f on each element, stopping at the first None *)
let rec iter_opt f = function
  | [] -> Some ()
  | x :: xs ->
    let* () = f x in
    iter_opt f xs

(* like iter_opt but also passes the 0-based index *)
let iteri_opt f xs =
  let rec go i = function
    | [] -> Some ()
    | x :: xs ->
      let* () = f i x in
      go (i + 1) xs
  in
  go 0 xs

type env = {
  buf      : Buffer.t;
  offsets  : (string, int) Hashtbl.t;
  mutable label_id : int;
  prefix    : string;
  ret_label : string;
  mutable loops : (string * string) list;
}

let emit env fmt =
  Printf.ksprintf
    (fun s ->
       Buffer.add_string env.buf s;
       Buffer.add_char env.buf '\n')
    fmt

let fresh_label env =
  let n = env.label_id in
  env.label_id <- n + 1;
  Printf.sprintf ".L%s_%d" env.prefix n

let push env = emit env "\tstr\tx0, [sp, #-16]!"
let pop env reg = emit env "\tldr\t%s, [sp], #16" reg

let align16 n = (n + 15) / 16 * 16

let assign_offsets env params body =
  let next = ref 0 in
  let alloc name =
    next := !next + 8;
    Hashtbl.replace env.offsets name (- !next)
  in
  List.iteri
    (fun i p ->
       if i < 8 then alloc p.p_name
       else Hashtbl.replace env.offsets p.p_name (16 + 8 * (i - 8)))
    params;
  let rec scan = function
    | Compound (ds, ss) ->
      List.iter (fun d -> alloc d.vd_name) ds;
      List.iter scan ss
    | If (_, t, e) ->
      let () = scan t in
      (match e with Some s -> scan s | None -> ())
    | While (_, b) -> scan b
    | Empty | Continue | Break | Return _ | Expr _ -> ()
  in
  let () = scan body in
  align16 !next

let offset_of env name =
  match Hashtbl.find_opt env.offsets name with
  | Some off -> Some off
  | None -> fail (Printf.sprintf "undefined variable %s" name)

let rec gen_expr env e =
  match e with
  | Int n ->
    let () = emit env "\tmov\tx0, #%d" n in
    Some ()
  | Var name ->
    let* off = offset_of env name in
    let () = emit env "\tldr\tx0, [x29, #%d]" off in
    Some ()
  | Assign (Var name, rhs) ->
    let* () = gen_expr env rhs in
    let* off = offset_of env name in
    let () = emit env "\tstr\tx0, [x29, #%d]" off in
    Some ()
  | Assign (_, _) ->
    fail "left-hand side of assignment must be a variable"
  | Unary (op, e) ->
    let* () = gen_expr env e in
    gen_unop env op
  | Binary (op, a, b) ->
    let* () = gen_expr env a in
    let () = push env in
    let* () = gen_expr env b in
    let () = pop env "x1" in
    gen_binop env op
  | Call (f, args) -> gen_call env f args

and gen_unop env op =
  match op with
  | Pos -> Some ()
  | Neg ->
    let () = emit env "\tneg\tx0, x0" in
    Some ()
  | BNot ->
    let () = emit env "\tmvn\tx0, x0" in
    Some ()
  | LNot ->
    let () = emit env "\tcmp\tx0, #0" in
    let () = emit env "\tcset\tx0, eq" in
    Some ()

and gen_binop env op =
  match op with
  | Add ->
    let () = emit env "\tadd\tx0, x1, x0" in
    Some ()
  | Sub ->
    let () = emit env "\tsub\tx0, x1, x0" in
    Some ()
  | Mul ->
    let () = emit env "\tmul\tx0, x1, x0" in
    Some ()
  | Div ->
    let () = emit env "\tsdiv\tx0, x1, x0" in
    Some ()
  | Mod ->
    let () = emit env "\tsdiv\tx2, x1, x0" in
    let () = emit env "\tmsub\tx0, x2, x0, x1" in
    Some ()
  | Lt -> cmp_set env "lt"
  | Gt -> cmp_set env "gt"
  | Le -> cmp_set env "le"
  | Ge -> cmp_set env "ge"
  | Eq -> cmp_set env "eq"
  | Ne -> cmp_set env "ne"

and cmp_set env cond =
  let () = emit env "\tcmp\tx1, x0" in
  let () = emit env "\tcset\tx0, %s" cond in
  Some ()

and gen_call env f args =
  let args = Array.of_list args in
  let n = Array.length args in
  let k = min n 8 in
  let nstack = n - k in
  let area = align16 (8 * nstack) in
  let () = if area > 0 then emit env "\tsub\tsp, sp, #%d" area in
  let rec push_args i =
    if i >= k then Some ()
    else
      let* () = gen_expr env args.(i) in
      let () = push env in
      push_args (i + 1)
  in
  let* () = push_args 0 in
  let rec stack_args i =
    if i >= n then Some ()
    else
      let* () = gen_expr env args.(i) in
      let () = emit env "\tstr\tx0, [sp, #%d]" (16 * k + 8 * (i - 8)) in
      stack_args (i + 1)
  in
  let* () = stack_args 8 in
  let () =
    for i = 0 to k - 1 do
      emit env "\tldr\tx%d, [sp, #%d]" i (16 * (k - 1 - i))
    done
  in
  let () = if k > 0 then emit env "\tadd\tsp, sp, #%d" (16 * k) in
  let () = emit env "\tbl\t%s" f in
  let () = if area > 0 then emit env "\tadd\tsp, sp, #%d" area in
  Some ()

let rec gen_stmt env s =
  match s with
  | Empty -> Some ()
  | Expr e -> gen_expr env e
  | Return e ->
    let* () = gen_expr env e in
    let () = emit env "\tb\t%s" env.ret_label in
    Some ()
  | Compound (_, ss) -> iter_opt (gen_stmt env) ss
  | If (cond, then_s, else_opt) ->
    let l_else = fresh_label env in
    let l_end = fresh_label env in
    let* () = gen_expr env cond in
    let () = emit env "\tcmp\tx0, #0" in
    let () = emit env "\tbeq\t%s" l_else in
    let* () = gen_stmt env then_s in
    let () = emit env "\tb\t%s" l_end in
    let () = emit env "%s:" l_else in
    let* () =
      match else_opt with
      | Some s -> gen_stmt env s
      | None -> Some ()
    in
    let () = emit env "%s:" l_end in
    Some ()
  | While (cond, body) ->
    let l_top = fresh_label env in
    let l_end = fresh_label env in
    let () = emit env "%s:" l_top in
    let* () = gen_expr env cond in
    let () = emit env "\tcmp\tx0, #0" in
    let () = emit env "\tbeq\t%s" l_end in
    let () = env.loops <- (l_end, l_top) :: env.loops in
    let* () = gen_stmt env body in
    let () = env.loops <- List.tl env.loops in
    let () = emit env "\tb\t%s" l_top in
    let () = emit env "%s:" l_end in
    Some ()
  | Break ->
    (match env.loops with
     | (brk, _) :: _ ->
       let () = emit env "\tb\t%s" brk in
       Some ()
     | [] -> fail "break outside of loop")
  | Continue ->
    (match env.loops with
     | (_, cont) :: _ ->
       let () = emit env "\tb\t%s" cont in
       Some ()
     | [] -> fail "continue outside of loop")

let gen_fun buf (f : fun_definition) =
  let env = {
    buf;
    offsets = Hashtbl.create 16;
    label_id = 0;
    prefix = f.f_name;
    ret_label = Printf.sprintf ".Lret_%s" f.f_name;
    loops = [];
  } in
  let frame = assign_offsets env f.f_params f.f_body in
  let () = emit env "\t.text" in
  let () = emit env "\t.global\t%s" f.f_name in
  let () = emit env "%s:" f.f_name in
  let () = emit env "\tstp\tx29, x30, [sp, #-16]!" in
  let () = emit env "\tmov\tx29, sp" in
  let () = if frame > 0 then emit env "\tsub\tsp, sp, #%d" frame in
  let store_param i p =
    if i < 8 then
      let* off = offset_of env p.p_name in
      let () = emit env "\tstr\tx%d, [x29, #%d]" i off in
      Some ()
    else Some ()
  in
  let* () = iteri_opt store_param f.f_params in
  let* () = gen_stmt env f.f_body in
  let () = emit env "\tmov\tx0, #0" in
  let () = emit env "%s:" env.ret_label in
  let () = emit env "\tmov\tsp, x29" in
  let () = emit env "\tldp\tx29, x30, [sp], #16" in
  let () = emit env "\tret" in
  Some ()

let gen_program (prog : program) : string option =
  let buf = Buffer.create 1024 in
  let gen_one (FunDef f) =
    let* () = gen_fun buf f in
    let () = Buffer.add_char buf '\n' in
    Some ()
  in
  let* () = iter_opt gen_one prog in
  Some (Buffer.contents buf)

File: jupyter2/ps/minc/minc/rs/minc/src/codegen.rs
/** begin hidden */
use crate::ast::*;
use std::collections::HashMap;
use std::fmt::Write;

pub struct Codegen {
    buf: String,
    offsets: HashMap<String, i64>,
    label_id: i64,
    prefix: String,
    ret_label: String,
    loops: Vec<(String, String)>,
}

fn align16(n: i64) -> i64 {
    (n + 15) / 16 * 16
}

fn new_codegen() -> Codegen {
    Codegen {
        buf: String::new(),
        offsets: HashMap::new(),
        label_id: 0,
        prefix: String::new(),
        ret_label: String::new(),
        loops: Vec::new(),
    }
}

fn emit(g: &mut Codegen, s: &str) {
    g.buf.push_str(s);
    g.buf.push('\n');
}

fn fresh_label(g: &mut Codegen) -> String {
    let l = format!(".L{}_{}", g.prefix, g.label_id);
    g.label_id += 1;
    l
}

fn push(g: &mut Codegen) {
    emit(g, "\tstr\tx0, [sp, #-16]!");
}
fn pop(g: &mut Codegen, reg: &str) {
    emit(g, &format!("\tldr\t{}, [sp], #16", reg));
}

fn assign_offsets(g: &mut Codegen, params: &[Param], body: &Compound) -> i64 {
    let mut next: i64 = 0;
    for (i, p) in params.iter().enumerate() {
        if i < 8 {
            next += 8;
            g.offsets.insert(p.name.clone(), -next);
        } else {
            g.offsets
                .insert(p.name.clone(), 16 + 8 * (i as i64 - 8));
        }
    }
    scan_stmt_compound(g, body, &mut next);
    align16(next)
}

fn scan_stmt_compound(g: &mut Codegen, c: &Compound, next: &mut i64) {
    for d in &c.decls {
        *next += 8;
        g.offsets.insert(d.name.clone(), -*next);
    }
    for s in &c.stmts {
        scan_stmt(g, s, next);
    }
}

fn scan_stmt(g: &mut Codegen, s: &Stmt, next: &mut i64) {
    match s {
        Stmt::Compound(c) => scan_stmt_compound(g, c, next),
        Stmt::If(_, then, els) => {
            scan_stmt(g, then, next);
            if let Some(e) = els {
                scan_stmt(g, e, next);
            }
        }
        Stmt::While(_, body) => scan_stmt(g, body, next),
        _ => {}
    }
}

fn offset_of(g: &Codegen, name: &str) -> Result<i64, String> {
    g.offsets
        .get(name)
        .copied()
        .ok_or_else(|| format!("codegen error: undefined variable {}", name))
}

fn gen_expr(g: &mut Codegen, e: &Expr) -> Result<(), String> {
    match e {
        Expr::IntLit(v) => emit(g, &format!("\tmov\tx0, #{}", v)),
        Expr::VarRef(name) => {
            let off = offset_of(g, name)?;
            emit(g, &format!("\tldr\tx0, [x29, #{}]", off));
        }
        Expr::Assign(lhs, rhs) => {
            let name = match lhs.as_ref() {
                Expr::VarRef(n) => n.clone(),
                _ => {
                    return Err(
                        "codegen error: left-hand side of assignment must be a variable"
                            .to_string(),
                    );
                }
            };
            gen_expr(g, rhs)?;
            let off = offset_of(g, &name)?;
            emit(g, &format!("\tstr\tx0, [x29, #{}]", off));
        }
        Expr::Unary(op, sub) => {
            gen_expr(g, sub)?;
            match op {
                UnaryOp::Pos => {}
                UnaryOp::Neg => emit(g, "\tneg\tx0, x0"),
                UnaryOp::BNot => emit(g, "\tmvn\tx0, x0"),
                UnaryOp::LNot => {
                    emit(g, "\tcmp\tx0, #0");
                    emit(g, "\tcset\tx0, eq");
                }
            }
        }
        Expr::Binary(op, l, r) => {
            gen_expr(g, l)?;
            push(g);
            gen_expr(g, r)?;
            pop(g, "x1");
            match op {
                BinaryOp::Add => emit(g, "\tadd\tx0, x1, x0"),
                BinaryOp::Sub => emit(g, "\tsub\tx0, x1, x0"),
                BinaryOp::Mul => emit(g, "\tmul\tx0, x1, x0"),
                BinaryOp::Div => emit(g, "\tsdiv\tx0, x1, x0"),
                BinaryOp::Mod => {
                    emit(g, "\tsdiv\tx2, x1, x0");
                    emit(g, "\tmsub\tx0, x2, x0, x1");
                }
                BinaryOp::Lt => cmp_set(g, "lt"),
                BinaryOp::Gt => cmp_set(g, "gt"),
                BinaryOp::Le => cmp_set(g, "le"),
                BinaryOp::Ge => cmp_set(g, "ge"),
                BinaryOp::Eq => cmp_set(g, "eq"),
                BinaryOp::Ne => cmp_set(g, "ne"),
            }
        }
        Expr::Call(name, args) => gen_call(g, name, args)?,
    }
    Ok(())
}

fn cmp_set(g: &mut Codegen, cond: &str) {
    emit(g, "\tcmp\tx1, x0");
    emit(g, &format!("\tcset\tx0, {}", cond));
}

fn gen_call(g: &mut Codegen, name: &str, args: &[Expr]) -> Result<(), String> {
    let n = args.len();
    let k = if n > 8 { 8 } else { n };
    let nstack = n - k;
    let area = align16(8 * nstack as i64);
    if area > 0 {
        emit(g, &format!("\tsub\tsp, sp, #{}", area));
    }
    for i in 0..k {
        gen_expr(g, &args[i])?;
        push(g);
    }
    for i in 8..n {
        gen_expr(g, &args[i])?;
        emit(g, &format!("\tstr\tx0, [sp, #{}]", 16 * k + 8 * (i - 8)));
    }
    for i in 0..k {
        emit(g, &format!("\tldr\tx{}, [sp, #{}]", i, 16 * (k - 1 - i)));
    }
    if k > 0 {
        emit(g, &format!("\tadd\tsp, sp, #{}", 16 * k));
    }
    emit(g, &format!("\tbl\t{}", name));
    if area > 0 {
        emit(g, &format!("\tadd\tsp, sp, #{}", area));
    }
    Ok(())
}

fn gen_stmt(g: &mut Codegen, s: &Stmt) -> Result<(), String> {
    match s {
        Stmt::Empty => {}
        Stmt::ExprStmt(e) => gen_expr(g, e)?,
        Stmt::Return(e) => {
            gen_expr(g, e)?;
            let rl = g.ret_label.clone();
            emit(g, &format!("\tb\t{}", rl));
        }
        Stmt::Compound(c) => {
            for x in &c.stmts {
                gen_stmt(g, x)?;
            }
        }
        Stmt::If(cond, then, els) => {
            let l_else = fresh_label(g);
            let l_end = fresh_label(g);
            gen_expr(g, cond)?;
            emit(g, "\tcmp\tx0, #0");
            emit(g, &format!("\tbeq\t{}", l_else));
            gen_stmt(g, then)?;
            emit(g, &format!("\tb\t{}", l_end));
            emit(g, &format!("{}:", l_else));
            if let Some(e) = els {
                gen_stmt(g, e)?;
            }
            emit(g, &format!("{}:", l_end));
        }
        Stmt::While(cond, body) => {
            let l_top = fresh_label(g);
            let l_end = fresh_label(g);
            emit(g, &format!("{}:", l_top));
            gen_expr(g, cond)?;
            emit(g, "\tcmp\tx0, #0");
            emit(g, &format!("\tbeq\t{}", l_end));
            g.loops.push((l_end.clone(), l_top.clone()));
            gen_stmt(g, body)?;
            g.loops.pop();
            emit(g, &format!("\tb\t{}", l_top));
            emit(g, &format!("{}:", l_end));
        }
        Stmt::Break => match g.loops.last() {
            Some((b, _)) => {
                let b = b.clone();
                emit(g, &format!("\tb\t{}", b));
            }
            None => return Err("codegen error: break outside of loop".to_string()),
        },
        Stmt::Continue => match g.loops.last() {
            Some((_, c)) => {
                let c = c.clone();
                emit(g, &format!("\tb\t{}", c));
            }
            None => return Err("codegen error: continue outside of loop".to_string()),
        },
    }
    Ok(())
}

fn gen_fun(g: &mut Codegen, f: &FunDef) -> Result<(), String> {
    g.offsets = HashMap::new();
    g.label_id = 0;
    g.prefix = f.name.clone();
    g.ret_label = format!(".Lret_{}", f.name);
    g.loops = Vec::new();
    let frame = assign_offsets(g, &f.params, &f.body);

    emit(g, "\t.text");
    emit(g, &format!("\t.global\t{}", f.name));
    emit(g, &format!("{}:", f.name));
    emit(g, "\tstp\tx29, x30, [sp, #-16]!");
    emit(g, "\tmov\tx29, sp");
    if frame > 0 {
        emit(g, &format!("\tsub\tsp, sp, #{}", frame));
    }
    for (i, p) in f.params.iter().enumerate() {
        if i < 8 {
            let off = offset_of(g, &p.name)?;
            emit(g, &format!("\tstr\tx{}, [x29, #{}]", i, off));
        }
    }
    let body = Stmt::Compound(f.body.clone());
    gen_stmt(g, &body)?;
    emit(g, "\tmov\tx0, #0");
    let rl = g.ret_label.clone();
    emit(g, &format!("{}:", rl));
    emit(g, "\tmov\tsp, x29");
    emit(g, "\tldp\tx29, x30, [sp], #16");
    emit(g, "\tret");
    Ok(())
}

pub fn gen_program(prog: &Program) -> Result<String, String> {
    let mut g = new_codegen();
    for f in &prog.funcs {
        gen_fun(&mut g, f)?;
        let _ = write!(g.buf, "\n");
    }
    Ok(g.buf)
}
/** end hidden */

/** begin skeleton */
use crate::ast::Program;
pub fn gen_program(prog: &Program) -> Result<String, String> {
    let _ = prog;
    Err("gen_program is not implemented yet".to_string())
}
/** end skeleton */
